### ISSUE TITLE
Authentication issues with 2.28

### ISSUE THREAD

**Issue by Neutrino-Sunset** [git-for-windows/git#2821]
- [x ] I was not able to find an [open](https://github.com/git-for-windows/git/issues?q=is%3Aopen) or [closed](https://github.com/git-for-windows/git/issues?q=is%3Aclosed) issue matching what I'm seeing

### Setup

 - Which version of Git for Windows are you using? Is it 32-bit or 64-bit?

```
git version 2.28.0.windows.1
cpu: x86_64
built from commit: 77982caf269b7ee713a76da2bcf260c34d3bf7a7
sizeof-long: 4
sizeof-size_t: 8
shell-path: /bin/sh
```

 - Which version of Windows are you running? Vista, 7, 8, 10? Is it 32-bit or 64-bit?

```
Microsoft Windows [Version 10.0.18362.1082]
```

 - What options did you set as part of the installation? Or did you choose the
   defaults?

```
Editor Option: VisualStudioCode
Custom Editor Path:
Path Option: Cmd
SSH Option: OpenSSH
Tortoise Option: false
CURL Option: WinSSL
CRLF Option: CRLFCommitAsIs
Bash Terminal Option: ConHost
Git Pull Behavior Option: Merge
Use Credential Manager: Enabled
Performance Tweaks FSCache: Enabled
Enable Symlinks: Disabled
Enable Pseudo Console Support: Disabled

```

 - Any other interesting things about your environment that might be related
   to the issue you're seeing?

Bog standard and up-to-date Win10 and vscode.

### Details

 - Which terminal/shell are you running Git from? e.g Bash/CMD/PowerShell/other

CMD and vscode integrated terminal

 - What commands did you run to trigger this issue? If you can provide a
   [Minimal, Complete, and Verifiable example](http://stackoverflow.com/help/mcve)
   this will help us understand the issue.

Created a new private repo on VisualStudio.com (Azure Devops).
Deleted all access tokens on VisualStudio.com itself and from local machine Control Panel -> Credential Manager

```
C:\Home\Development\Workspaces\Scratch>git clone https://neutrinosunset.visualstudio.com/Test1/_git/Test1
Cloning into 'Test1'...
fatal: unable to access 'https://neutrinosunset.visualstudio.com/Test1/_git/Test1/': schannel: failed to receive handshake, SSL/TLS connection failed
```
 - What did you expect to occur after running these commands?

For the Git Credential Manager to prompt me for some credentials, which would then be stored in the Windows Credential Manager and for the clone operation to succeed.

 - What actually happened instead?

```
fatal: unable to access 'https://neutrinosunset.visualstudio.com/Test1/_git/Test1/': schannel: failed to receive handshake, SSL/TLS connection failed
```

 - If the problem was occurring with a specific repository, can you provide the
   URL to that repository to help us with testing?

git clone https://neutrinosunset.visualstudio.com/Test1/_git/Test1

I've tried uninstalling all Git applications from the wor kstation, deleting all Git config files. Reinstalling 2.28 using every combination of the OpenSSL/Windows Certificate management and Windows Credential Manager/Core Credential Manager and nothing has worked.

In addition whenever I select the option to add the Unix tools to the system path that never works either, even after a system reboot.

**Comment by Neutrino-Sunset** [comment]
Following log shows an attempt to clone a private repo.

First attempt to clone using url https://neutrinosunset.visualstudio.com/EloquentJs/_git/EloquentJs fails.
Second attempt using url https://neutrino_sunset@neutrinosunset.visualstudio.com/EloquentJs/_git/EloquentJs also fails (note leading username)
Third attempt  using the same url as the first finally succeeds, pops up Credential Manager window, I enter username and password. Clone succeeds.

I then edit a file. Commit and push, and the push fails.

Just had the exact same thing happen with a brand new clean test repo too.


```
C:\Home\Development\Workspaces\Scratch>git clone https://neutrinosunset.visualstudio.com/EloquentJs/_git/EloquentJs
Cloning into 'EloquentJs'...
fatal: unable to access 'https://neutrinosunset.visualstudio.com/EloquentJs/_git/EloquentJs/': schannel: failed to receive handshake, SSL/TLS connection failed

C:\Home\Development\Workspaces\Scratch>git clone https://neutrino_sunset@neutrinosunset.visualstudio.com/EloquentJs/_git/EloquentJs
Cloning into 'EloquentJs'...
fatal: unable to access 'https://neutrinosunset.visualstudio.com/EloquentJs/_git/EloquentJs/': schannel: failed to receive handshake, SSL/TLS connection failed

C:\Home\Development\Workspaces\Scratch>git clone https://neutrino_sunset@neutrinosunset.visualstudio.com/EloquentJs/_git/EloquentJs
Cloning into 'EloquentJs'...
fatal: unable to access 'https://neutrinosunset.visualstudio.com/EloquentJs/_git/EloquentJs/': schannel: failed to receive handshake, SSL/TLS connection failed

C:\Home\Development\Workspaces\Scratch>git clone https://neutrinosunset.visualstudio.com/EloquentJs/_git/EloquentJs
Cloning into 'EloquentJs'...
remote: Azure Repos
remote: Found 130 objects to send. (161 ms)
Receiving objects: 100% (130/130), 33.33 KiB | 1.85 MiB/s, done.
Resolving deltas: 100% (41/41), done.

C:\Home\Development\Workspaces\Scratch>cd EloquentJs

C:\Home\Development\Workspaces\Scratch\EloquentJs>git status
On branch master
Your branch is up to date with 'origin/master'.

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
        modified:   main.js

no changes added to commit (use "git add" and/or "git commit -a")

C:\Home\Development\Workspaces\Scratch\EloquentJs>git add .

C:\Home\Development\Workspaces\Scratch\EloquentJs>git status
On branch master
Your branch is up to date with 'origin/master'.

Changes to be committed:
  (use "git restore --staged <file>..." to unstage)
        modified:   main.js


C:\Home\Development\Workspaces\Scratch\EloquentJs>git commit -m "Test"
[master 774e352] Test
 1 file changed, 2 insertions(+)

C:\Home\Development\Workspaces\Scratch\EloquentJs>git push
fatal: unable to access 'https://neutrinosunset.visualstudio.com/EloquentJs/_git/EloquentJs/': schannel: failed to receive handshake, SSL/TLS connection failed

```

**Comment by dscho** [comment]
Does it work when you use the OpenSSL backend, i.e. `git -c http.sslbackend=openssl push`?

**Comment by dscho** [comment]
(Another angle to try is to set `GCM_TRACE=1` and `GIT_TRACE_CURL=1` before trying to push again; Please review the output closely before pasting it anywhere as it might contain secrets.)

**Comment by Neutrino-Sunset** [comment]
I tried installing using the OpenSSL option, but that didn't work. I couldn't get it to work with https at all so in the end I used SSH which worked fine.

**Comment by dscho** [comment]
> [...] but that didn't work

Would you have detailed logs of that, maybe with `GIT_TRACE_CURL=1`? (I suspect that there might be proxy problems, but without logs I would be forced to guess.)

> in the end I used SSH which worked fine.

Oh, okay. If that is a satisfactory outcome for you, please close this ticket. Otherwise, I will try my best to assist in diagnosing this issue further.

**Comment by schwarrior** [comment]
This still happening in git version 2.30.1.windows.1. The SSH workaround is really inconvenient. Passphrase wont cache and must be re-entered on every remote operation.

**Comment by dscho** [comment]
Without anybody who experiences the issue acting on suggestions such as https://github.com/git-for-windows/git/issues/2821#issuecomment-699883073, there is little I can do here.

**Comment by schwarrior** [comment]
Okay I can authenticate with a Microsoft account again, even one with two factor authentication enabled.. Not sure which of these steps was the fix but here's what I did:
1) Uninstalled Git for Windows 2.30.1-64-bit
2) Opened Windows Credential Manager and deleted any saved credential over a year old, and anything newer that referenced GitHub, Visual Studio Online, Azure or the MS account I used to authenticate on these services.
3) Installed Git for Windows 2.32.0-64-bit
4) Was sure to select Git Credential Manager Core & OpenSSL during install
5) Made first connection to VisualStudio.com hosted GIT repo from a PowerShell terminal. A browser-esque login window popped up that took my Microsoft Account credentials and did 2FA with a text message. 
6) Subsequent remote ops (push / pull) from this terminal didn't require login. I assume this mean credentials were successfully cached by the Credential Manager.
7) Tested GIT remote access (push / pull) using integrated clients in Visual Studio 2019 and Visual Studio Code. No snags. No logins. Cached credentials worked.

**Comment by Neutrino-Sunset** [comment]
Just updated, created a new private Azure Devops project, all default settings.

```
C:\Home\Development\Workspaces>cmd.exe /c ver
Microsoft Windows [Version 10.0.18363.1556]
```

```
C:\Home\Development\Workspaces>git --version --build-options
git version 2.32.0.windows.1
cpu: x86_64
built from commit: 4c204998d0e156d13d81abe1d1963051b1418fc0
sizeof-long: 4
sizeof-size_t: 8
shell-path: /bin/sh
feature: fsmonitor--daemon
```

```
Editor Option: VisualStudioCode
Custom Editor Path: 
Default Branch Option: master
Path Option: Cmd
SSH Option: OpenSSH
Tortoise Option: false
CURL Option: OpenSSL
CRLF Option: CRLFCommitAsIs
Bash Terminal Option: ConHost
Git Pull Behavior Option: Merge
Use Credential Manager: Core
Performance Tweaks FSCache: Enabled
Enable Symlinks: Disabled
Enable Pseudo Console Support: Disabled
Enable FSMonitor: Disabled
```

Attempt to clone new repo.

**Expected:** To be prompted for account credentials and for the credential manager to create and store its own personal access token as described in the documentation here [Use Git Credential Manager Core to authenticate to Azure Repos](https://docs.microsoft.com/en-us/azure/devops/repos/git/set-up-credential-managers?toc=%2Fazure%2Fdevops%2Forganizations%2Ftoc.json&bc=%2Fazure%2Fdevops%2Forganizations%2Fbreadcrumb%2Ftoc.json&view=azure-devops#using-the-git-credential-manager)

**Result:** The following error.

```
C:\Home\Development\Workspaces>set GCM_TRACE=1

C:\Home\Development\Workspaces>set GIT_TRACE_CURL=1

C:\Home\Development\Workspaces>git clone https://neutrinosunset.visualstudio.com/Test3/_git/Test3
Cloning into 'Test3'...
20:45:45.088554 http.c:774              == Info: Couldn't find host neutrinosunset.visualstudio.com in the (nil) file; using defaults
20:45:45.097552 http.c:774              == Info:   Trying 2620:1ec:21::18:443...
20:45:45.113549 http.c:774              == Info: Connected to neutrinosunset.visualstudio.com (2620:1ec:21::18) port 443 (#0)
20:45:45.113549 http.c:774              == Info: ALPN, offering h2
20:45:45.113549 http.c:774              == Info: ALPN, offering http/1.1
20:45:45.126556 http.c:774              == Info: successfully set certificate verify locations:
20:45:45.126556 http.c:774              == Info:  CAfile: C:/Programs/Git/mingw64/ssl/certs/ca-bundle.crt
20:45:45.126556 http.c:774              == Info:  CApath: none
20:45:45.127553 http.c:747              => Send SSL data, 0000000005 bytes (0x00000005)
20:45:45.127553 http.c:762              => Send SSL data: .....
20:45:45.127553 http.c:774              == Info: TLSv1.3 (OUT), TLS handshake, Client hello (1):
20:45:45.127553 http.c:747              => Send SSL data, 0000000512 bytes (0x00000200)
20:45:45.127553 http.c:762              => Send SSL data: ......p....%..&....cY.k.U....*....B..% uK.Y..?..j......../<y
20:45:45.127553 http.c:762              => Send SSL data: .../X.[..qC.>.......,.0.........+./...$.(.k.#.'.g.....9.....
20:45:45.127553 http.c:762              => Send SSL data: 3.....=.<.5./.....u...$."...neutrinosunset.visualstudio.com.
20:45:45.127553 http.c:762              => Send SSL data: .......................3t.........h2.http/1.1.........1.....
20:45:45.127553 http.c:762              => Send SSL data: 0.................................................+.........
20:45:45.127553 http.c:762              => Send SSL data: ...-.....3.&.$... .;.o..b..N..v.....GZ..o..Iyh.g.D..........
20:45:45.127553 http.c:762              => Send SSL data: ............................................................
20:45:45.127553 http.c:762              => Send SSL data: ............................................................
20:45:45.127553 http.c:762              => Send SSL data: ................................
20:45:45.127553 http.c:774              == Info: OpenSSL SSL_connect: Connection was reset in connection to neutrinosunset.visualstudio.com:443
F'
20:45:45.127553 http.c:774              == Info: Closing connection 0
fatal: unable to access 'https://neutrinosunset.visualstudio.com/Test3/_git/Test3/': OpenSSL SSL_connect: Connection was reset in connection to neutrinosunset.visualstudio.com:443
```

**Comment by dscho** [comment]
A colleague of mine had a hunch that this might be caused by the use of IPv6. Could you try cloning with the [`--ipv4`](https://git-scm.com/docs/git-fetch#Documentation/git-fetch.txt--4) option?

**Comment by Neutrino-Sunset** [comment]
That resulted in some progress. On the first attempt that took me to the git credentials prompt on the command line, but entering my Microsoft account credentials there failed to authenticate. What should happen is that the credential manager should prompt me for me Microsoft account credentals and use those to generate a PAT token to use for the Git authentication.

However on the second attempt it worked and the credential manager popup appeared and accepted my credentials and then cloned the repo.

**Comment by dscho** [comment]
> On the first attempt that took me to the git credentials prompt on the command line, but entering my Microsoft account credentials there failed to authenticate.

This looks like a bug in Git Credential Manager Core to me. Or maybe it is Git that does not even give the credential manager a chance to say "oh, my credentials did not work? Let me try again?", or something.

@mjcheetham would you have any insight about this behavior?

**Comment by dscho** [comment]
This vaguely looks like network issues. Do you have a proxy?

A big hammer to diagnose this better might be to install Fiddler (but you would need to analyze the logs largely by yourself as it contains plenty of secrets, I'd wager).

**Comment by mjcheetham** [comment]
@Neutrino-Sunset, have you made any custom TLS setting changes in Windows? It's possible the connection is being reset by Azure DevOps because of some TLS handshake problem.

The fact this error is happening intermittently in the network stacks of both GCM (.NET Framework) _and_ Git (cURL) tells me this is an OS networking issue.

In fact, I notice from your earlier `GIT_TRACE_CURL` logs that your machine is using TLS 1.3:
```
...
20:45:45.127553 http.c:774              == Info: TLSv1.3 (OUT), TLS handshake, Client hello (1):
...
```

You're also on Windows 10 version 1903 (build 18362), which has an incomplete implementation of TLS 1.3.

```
Microsoft Windows [Version 10.0.18362.1082]
```

See this [Microsoft article about TLS 1.3 support](https://devblogs.microsoft.com/premier-developer/microsoft-tls-1-3-support-reference/), and specifically this part:

> **Will TLS 1.3 be supported in Windows 10 and Server?**
>
> TLS 1.3 is also supported on Windows 1903 as of release of this article for testing purposes only, not production environment.

**Comment by Neutrino-Sunset** [comment]
I haven't made any changes to the networking on this machine at all.

**Comment by rimrul** [comment]
Maybe a group policy enabling or even enforcing TLS 1.3?

**Comment by Neutrino-Sunset** [comment]
It's a private laptop not on a domain so no network policy is being applied.

I've checked the inetcpl.cpl settings and TLS 1.3 is not being applied there nor in the registry.

If anything is choosing to use TLS 1.3 by itself I would guess the culprit would be OpenSSL which has support for it. I'll reinstall using WinSSL and see if that helps.

**Comment by Neutrino-Sunset** [comment]
I reinstalled Git selecting WinSSL.

Using the `--ipv4` option I then cloned my repo and pushed a change to it. Three times in a row I did this without any problems, deleting the repo and credentials on Windows and on Azure each time. It worked every time.

I then did the same thing all over again without the `--ipv4` option. It still worked.

I then reinstalled Git selecting the OpenSLL option. On the second attempt to clone the repo it failed with the same connection reset error as earlier.

```
20:45:45.127553 http.c:774              == Info: OpenSSL SSL_connect: Connection was reset in connection to neutrinosunset.visualstudio.com:443
F'
20:45:45.127553 http.c:774              == Info: Closing connection 0
fatal: unable to access 'https://neutrinosunset.visualstudio.com/Test3/_git/Test3/': OpenSSL SSL_connect: Connection was reset in connection to neutrinosunset.visualstudio.com:443
```

I then reintalled using WinSSL and performed three more complete cycles of clone, push, delete reop, clear credentials, with no errors.

It seems to me that whatever my original problem was with my initial WinSLL install causing the schannel error is fixed in this current build. However the OpenSSL implementation in the current build seems to have some problems on Win 10.
